arelle/formula/FactAspectsCache.py

```diff
@@ -10,24 +10,60 @@
 
 
 class FactAspectsCache:
-    def __init__(self) -> None:
-        self._matchingAspects: defaultdict[ModelFact, defaultdict[ModelFact, defaultdict[int | QName, bool | None]]] = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: None)))
+    def __init__(self, maxSize: int) -> None:
+        self._size = 0
+        self._maxSize = maxSize if maxSize >= 0 else float("inf")
+        self._maxSizePreviouslyReached = False
+        self._prioritizedAspects: set[int | QName] = set()
+        self._matchingAspects: defaultdict[
+            ModelFact,
+            defaultdict[
+                ModelFact,
+                defaultdict[int | QName, bool | None]
+            ]
+        ] = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: None)))
 
-    def evaluations(self, fact1: ModelFact, fact2: ModelFact) -> defaultdict[int | QName, bool | None]:
-        return self._matchingAspects[fact1][fact2]
+    @property
+    def prioritizedAspects(self) -> set[int | QName]:
+        return self._prioritizedAspects
+
+    def evaluations(self, fact1: ModelFact, fact2: ModelFact) -> defaultdict[int | QName, bool | None] | None:
+        if not self._maxSizePreviouslyReached:
+            if fact1 in self._matchingAspects and fact2 in self._matchingAspects[fact1]:
```

Comment:
Out of curiosity, did you consider adding a unique fact integer ID to each fact?  The dict key could be a (min(id1,id2),max(id1,id2),aspects) tuple.

Reply:
I experimented with tuple keys (it wasn't sorted and still performed a double lookup), but that had worse performance. I'll run another experiment with this specifically and open a followup PR if it's better

Reply:
Interesting.  My intuition was that a single dict lookup would be better than multiple, but perhaps tuple overhead makes that not worth it.  I suppose even with multi-level dict, having an orderable id so you know which order to do the dict lookup might help, but perhaps not much.